File: sdk/go/client/concept_browser.go
package client

import (
	"context"
	"fmt"
)

// DefaultConceptBrowsePageSize is the keyset page size BrowseConcept uses
// when the caller doesn't specify one. A concept registry can hold far more
// than the engine's implicit unmarked-list backstop (50 rows, the
// MEMORY_ENGINE_DEFAULT_LIST_CAP), so the concept-browse query MUST declare
// sort + paginate to opt into a keyset window + a continuation cursor --
// otherwise it silently truncates at the backstop with no way to page past
// it (memql#2008). 200 is a single comfortable page for the cockpit's row
// list; larger registries continue via the cursor.
const DefaultConceptBrowsePageSize = 200

// BrowseConcept returns every row stored under the given concept id.
//
// Bypasses the named-primitive surface intentionally. The Cockpit's
// concept-browser tab is the canonical caller: it iterates the
// concept registry from ListConcepts and lets an operator click into
// any concept's rows without knowing its name at compile time. That
// use case is concept-name-agnostic by definition, so there's no
// equivalent named primitive in the DSL tree.
//
// Every OTHER caller should reach for a typed generated method on
// QueryClient. Direct concept browsing is reserved for admin /
// debug surfaces -- the named-primitive rule (see sdk/go/CLAUDE.md)
// applies to product code without exception.
//
// Deprecated for large registries: returns only the first
// DefaultConceptBrowsePageSize rows with no way to continue. Prefer
// BrowseConceptPage, which exposes the keyset cursor so the caller can walk
// the whole concept (memql#2008). Retained because the single-page form is
// still the common case for small concepts.
func (qc *QueryClient) BrowseConcept(ctx context.Context, conceptId string) (*Result, error) {
	page, err := qc.BrowseConceptPage(ctx, conceptId, "", 0)
	if err != nil {
		return nil, err
	}
	// Re-wrap the page rows in a bundle envelope so the legacy Result
	// accessors (RawNodes / Rows) keep working unchanged.
	nodes := make([]any, 0, len(page.Rows))
	for _, r := range page.Rows {
		nodes = append(nodes, map[string]any(r))
	}
	return &Result{payload: map[string]any{
		"bundle": map[string]any{"nodes": nodes},
	}}, nil
}

// BrowseConceptPage is the keyset-paginated form of BrowseConcept. It returns
// one page of rows under the given concept id, oldest-first (createdAt ASC,
// with the id ASC tie-breaker the engine appends under equal createdAt), plus
// the continuation cursor for the next page.
//
// cursor   -- opaque continuation token from a prior page's
//
//	PageResult.NextCursor; "" for the first page.
//
// pageSize -- rows per page; <= 0 uses DefaultConceptBrowsePageSize.
//
// Walking the full registry:
//
//	cursor := ""
//	for {
//	    page, err := qc.BrowseConceptPage(ctx, id, cursor, 0)
//	    if err != nil { return err }
//	    rows = append(rows, page.Rows...)
//	    cursor = page.NextCursor
//	    if cursor == "" { break } // exhausted
//	}
//
// PageResult.Rows preserves the full nested wire shape (payload / metadata /
// provenance / intrinsics) for the bundle path, exactly like
// Result.RawNodes() -- so the cockpit's generic concept inspector renders
// unchanged.
//
// The query declares `sort` + `paginate` so it opts into a bounded keyset
// window + a next-page cursor instead of hitting the engine's implicit
// unmarked-list backstop (the MEMORY_ENGINE_DEFAULT_LIST_CAP, 50 rows) with
// no continuation (memql#2008). The cursor rides ExecuteQueryMsg.cursor and
// is bound to this query's sort ordering; it resolves on any replica.
func (qc *QueryClient) BrowseConceptPage(ctx context.Context, conceptId, cursor string, pageSize int) (*PageResult, error) {
	if conceptId == "" {
		return nil, fmt.Errorf("BrowseConcept: conceptId is required")
	}
	if pageSize <= 0 {
		pageSize = DefaultConceptBrowsePageSize
	}
	// sort(paginate(concept==X, N), "createdAt", "asc") -- the canonical
	// keyset-eligible directive chain (leading createdAt sort + a paginate
	// window), mirroring keysetPageQuery in the engine's keyset DB test. The
	// engine appends `id ASC` as the tie-breaker under equal createdAt; see
	// component/memql/cursor.go.
	query := fmt.Sprintf(`sort(paginate(concept==%s, %d), "createdAt", "asc")`, conceptId, pageSize)
	page, err := qc.ExecutePaginated(ctx, query, cursor)
	if err != nil {
		return nil, fmt.Errorf("BrowseConcept(%s): %w", conceptId, err)
	}
	return page, nil
}

// GetRowByConceptAndId returns the single row matching (conceptId,
// rowId). Same admin-surface caveat as BrowseConcept -- product code
// should use a typed lookup primitive (userById, agentById,
// etc.) instead.
func (qc *QueryClient) GetRowByConceptAndId(ctx context.Context, conceptId, rowId string) (*Result, error) {
	if conceptId == "" {
		return nil, fmt.Errorf("GetRowByConceptAndId: conceptId is required")
	}
	if rowId == "" {
		return nil, fmt.Errorf("GetRowByConceptAndId: rowId is required")
	}
	payload, err := qc.executeRaw(ctx, fmt.Sprintf("concept==%s;id==%s", conceptId, rowId))
	if err != nil {
		return nil, fmt.Errorf("GetRowByConceptAndId(%s, %s): %w", conceptId, rowId, err)
	}
	return &Result{payload: payload}, nil
}

File: sdk/ts/src/client/conceptBrowser.ts
// Generic concept browsing -- the TypeScript mirror of sdk/go's
// concept_browser.go.
//
// This bypasses the named-primitive surface deliberately, and it is the ONLY
// sanctioned reason to do so. A concept browser iterates the registry from
// listConcepts and lets an operator click into any concept's rows without
// knowing its name at compile time; that use case is concept-name-agnostic by
// definition, so no equivalent named primitive can exist. Every other caller
// uses a generated typed method.

import type { QueryClient, QueryCallOptions } from "./query.js";
import type { Row } from "./types.js";

// A concept registry can hold far more than the engine's implicit
// unmarked-list backstop (MEMQL_MEMORY_ENGINE_DEFAULT_LIST_CAP, 50 rows), so the
// browse query MUST declare sort + paginate to opt into a keyset window and a
// continuation cursor -- otherwise it silently truncates at the backstop with
// no way to page past it (memql#2008).
export const DEFAULT_CONCEPT_BROWSE_PAGE_SIZE = 200;

export interface ConceptPage {
  // rows preserve the full nested wire shape (payload / metadata / provenance
  // / intrinsics), exactly like Result.rawNodes(). A generic inspector needs
  // the intrinsics that flattening would drop.
  rows: Row[];
  // nextCursor is the opaque continuation token for the following page, or ""
  // when the set is exhausted. It is bound to this query's sort ordering and
  // carries no server session state, so it resolves on any replica.
  nextCursor: string;
}

export interface ConceptBrowseOptions extends QueryCallOptions {
  /**
   * Walk direction. The default "asc" is the canonical browse ordering the
   * concept workspace depends on (oldest first, live arrivals banded apart).
   * "desc" exists for surfaces whose question is "what happened RECENTLY" --
   * the observability drill-in reads codeMetric windows newest-first
   * (memql#4192) -- and mints the same keyset cursor bound to its own
   * ordering. Additive: nothing changes for callers that do not pass it.
   */
  order?: "asc" | "desc";
  pageSize?: number;
}

export async function browseConceptPage(
  query: QueryClient,
  conceptId: string,
  opts: ConceptBrowseOptions = {},
): Promise<ConceptPage> {
  if (conceptId === "") {
    throw new Error("browseConceptPage: conceptId is required");
  }
  // Split off the browse-only knob and forward EVERYTHING else verbatim.
  // Hand-copying `cursor` + `signal` was correct for today's
  // QueryCallOptions and silently wrong the moment a field is added to it --
  // the new option would compile fine here and never reach the query.
  const { pageSize: requestedPageSize, order, ...callOpts } = opts;
  const pageSize =
    requestedPageSize !== undefined && requestedPageSize > 0
      ? requestedPageSize
      : DEFAULT_CONCEPT_BROWSE_PAGE_SIZE;

  // sort(paginate(concept==X, N), "createdAt", "asc") -- the canonical
  // keyset-eligible directive chain (leading createdAt sort plus a paginate
  // window). The engine appends `id ASC` as the tie-breaker under equal
  // createdAt.
  const direction = order === "desc" ? "desc" : "asc";
  const call = `sort(paginate(concept==${conceptId}, ${pageSize}), "createdAt", "${direction}")`;

  const result = await query.executeNamed("conceptBrowse", call, callOpts);

  return {
    rows: result.rawNodes(),
    nextCursor: result.meta()?.cursor ?? "",
  };
}

export async function getRowByConceptAndId(
  query: QueryClient,
  conceptId: string,
  rowId: string,
  opts: QueryCallOptions = {},
): Promise<Row | null> {
  if (conceptId === "") {
    throw new Error("getRowByConceptAndId: conceptId is required");
  }
  if (rowId === "") {
    throw new Error("getRowByConceptAndId: rowId is required");
  }
  const result = await query.executeNamed(
    "conceptRow",
    `concept==${conceptId} && id==${rowId}`,
    opts,
  );
  const nodes = result.rawNodes();
  return nodes.length > 0 ? (nodes[0] ?? null) : null;
}

/**
 * Count the rows of a concept the CALLER may see.
 *
 * The engine's `count` directive (memql#1730) computes this server-side through
 * the same pipeline a normal query uses -- deduped, latest-version,
 * post-filtered, and under the caller's own per-row authz -- and returns a
 * `{count: N}` envelope instead of materializing rows. A raw SQL COUNT(*) would
 * over-count under the time-series versioning model (many versions per id) and
 * skip the in-process folds, which is why this is a directive rather than a
 * cheaper query.
 *
 * WHY THIS EXISTS SEPARATELY FROM browseConceptPage. A surface wanting "how
 * many customers are there" used to fetch a page of rows and count what came
 * back, which answers a different question: it caps at the page size and has to
 * render "100+". Two calls -- this for the number, a bounded page for the rows
 * a surface actually shows -- is both cheaper and honest (memql#4263).
 *
 * The count is PER-CALLER by construction. A reader's count is the rows a
 * reader may see, and that is the correct number to show the person looking at
 * it; do not reach for a privileged count to make it look bigger.
 */
export async function countConcept(
  query: QueryClient,
  conceptId: string,
  opts: QueryCallOptions = {},
): Promise<number> {
  if (conceptId === "") {
    throw new Error("countConcept: conceptId is required");
  }
  const result = await query.executeNamed(
    "conceptCount",
    `count(concept==${conceptId})`,
    opts,
  );

  // The engine sets ResultMeta.count AND returns a {count: N} output row. Read
  // the meta first (it is the typed field), and fall back to the row for a
  // node that answered with only the envelope.
  const meta = result.meta();
  if (meta?.count !== undefined && meta.count !== null) {
    const n = typeof meta.count === "string" ? Number(meta.count) : meta.count;
    if (Number.isFinite(n) && n >= 0) return n;
  }
  const row = result.single();
  const raw = row?.["count"];
  const n = typeof raw === "string" ? Number(raw) : typeof raw === "number" ? raw : NaN;
  if (!Number.isFinite(n) || n < 0) {
    // -1 is the engine's "unknown count" sentinel (component/memql/result.go).
    // Reporting it as a number would put "-1 people" on a console; refusing is
    // what lets the caller render its own "could not read" state.
    throw new Error(`countConcept: ${conceptId} returned no usable count`);
  }
  return n;
}
